### ISSUE TITLE
v0.9 API

### ISSUE THREAD

**Issue by stoically** [stoically/syn-rsx#26]
Follow-up from https://github.com/stoically/syn-rsx/pull/23

Issue for general API discussion of the upcoming v0.9 release, which will have a rather large breaking change upcoming: `Node` will be converted to an `enum`, and the node tree leafs are `Node` enums as well. This will come with additional type safety, but makes the nodes a bit harder to access in terms of API. Feel free to voice concerns or ideas.

**Should `NodeValueExpr` implement `Deref`? Is it a newtype or a smart pointer?**

After reading up a bit on what the book says about newtypes and smart pointers ([here](https://doc.rust-lang.org/book/ch19-04-advanced-types.html?highlight=newtype#using-the-newtype-pattern-for-type-safety-and-abstraction), [here](https://doc.rust-lang.org/book/ch19-03-advanced-traits.html#using-the-newtype-pattern-to-implement-external-traits-on-external-types) and [here](https://doc.rust-lang.org/book/ch15-02-deref.html#treating-smart-pointers-like-regular-references-with-the-deref-trait)), I'm leaning towards `NodeValueExpr` being a smart pointer. Reason being: While it doesn't hold extra metadata, its main purpose is granting access to the type it is encapsulating and adding extra capabilities via a multitude of traits. Furthermore consumers mainly want work with the encapsulated type, so having to do explicit method calls just to get the `Expr` instead of being able to leverage reference coercion seems like the wrong move.

My understanding regarding the newtype patterns is that the main use case is additional abstraction over encapsulated types – which also should protect internal type state from undesired modifications –, or specifically to work around the orphan rule. Both isn't what `NodeValueExpr` is doing/providing.

**Comment by Zizico2** [comment]
Ok, fair enough. I'm still on the fence about it, but I guess newtype vs smart pointers is not really a black vs white issue, it's more nuanced than that. And I like the renaming. `NodeValue`, to me, seemed like a newtype. `NodeValueExpr` makes it seem like it's an `Expr` (and it is), just with extra goodies.

**Comment by stoically** [comment]
> but I guess newtype vs smart pointers is not really a black vs white issue, it's more nuanced than that. 

Yeah, I'd agree.

> NodeValueExpr makes it seem like it's an Expr (and it is), just with extra goodies.

That was the intention, glad it worked out. Also considered `Value`, `ValueExpr` and just `Expr`.

**Comment by stoically** [comment]
Something I realized: The `impl TryFrom<NodeValue> for String` isn't really discoverable in the rust docs?! It's really only discoverable through the examples. I wonder what a good way to expose that bit of information would be? Just some hand-written explanation in `NodeValue`s doc comment? :thinking:

**Comment by Zizico2** [comment]
Is it implemented?

Cuz

```
source
impl TryFrom<&NodeValueExpr> for String
type Error = Error
The type returned in the event of a conversion error.
source
fn try_from(value: &NodeValueExpr) -> Result<Self, Self::Error>
Performs the conversion.
```

this one shows up.

**Comment by stoically** [comment]
Yeah, `TryFrom<&NodeValueExpr>, but still, it isn't discoverable in the rustdocs, or am I blind? :thinking:

**Comment by Zizico2** [comment]
`TryFrom<&NodeValueExpr>` shows up
`TryFrom<NodeValueExpr>` doesn't

**Comment by Zizico2** [comment]
You're looking at the alpha docs right?

**Comment by stoically** [comment]
Oooh, I'm blind, it's there https://docs.rs/syn-rsx/0.9.0-alpha.1/syn_rsx/struct.NodeValueExpr.html#impl-TryFrom%3C%26NodeValueExpr%3E-for-String, right. Then it's fine :D

**Comment by stoically** [comment]
Release scheduled for 10.11.2022.

**Comment by gbj** [comment]
I know the issue is closed, but just wanted to chime in on this and say thanks for the new API -- having just reimplemented our view macro using it and I can say it's much nicer to have the type safe enum variants than the constant .name_as_string().unwrap() and whatnot before. Great update!
